drivers/watchdog/wdt_nrfx.c

```diff
@@ -183,42 +185,47 @@ static void wdt_event_handler(const struct device *dev, nrf_wdt_event_t event_ty
 		requests &= ~BIT(i);
 	}
 }
+#endif
 
 #define WDT(idx) DT_NODELABEL(wdt##idx)
 
-#define WDT_NRFX_WDT_DEVICE(idx)					       \
-	static void wdt_##idx##_event_handler(nrf_wdt_event_t event_type,      \
-					      uint32_t requests,	       \
-					      void *p_context)		       \
-	{								       \
-		wdt_event_handler(DEVICE_DT_GET(WDT(idx)), event_type,         \
-				  requests, p_context);			       \
-	}								       \
-	static int wdt_##idx##_init(const struct device *dev)		       \
-	{								       \
-		const struct wdt_nrfx_config *config = dev->config;	       \
-		nrfx_err_t err_code;					       \
-		IRQ_CONNECT(DT_IRQN(WDT(idx)), DT_IRQ(WDT(idx), priority),     \
-			    nrfx_isr, nrfx_wdt_##idx##_irq_handler, 0);	       \
-		err_code = nrfx_wdt_init(&config->wdt,			       \
-					 NULL,				       \
-					 wdt_##idx##_event_handler,	       \
-					 NULL);				       \
-		if (err_code != NRFX_SUCCESS) {				       \
-			return -EBUSY;					       \
-		}							       \
-		return 0;						       \
-	}								       \
-	static struct wdt_nrfx_data wdt_##idx##_data;			       \
-	static const struct wdt_nrfx_config wdt_##idx##z_config = {	       \
-		.wdt = NRFX_WDT_INSTANCE(idx),				       \
-	};								       \
-	DEVICE_DT_DEFINE(WDT(idx),					       \
-			    wdt_##idx##_init,				       \
-			    NULL,					       \
-			    &wdt_##idx##_data,				       \
-			    &wdt_##idx##z_config,			       \
-			    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,  \
+#define WDT_NRFX_WDT_DEVICE(idx)							       \
+	COND_CODE_0(IS_ENABLED(CONFIG_WDT_NRFX_NO_IRQ), (				       \
+	static void wdt_##idx##_event_handler(nrf_wdt_event_t event_type,		       \
+					      uint32_t requests,			       \
+					      void *p_context)				       \
+	{										       \
+		wdt_event_handler(DEVICE_DT_GET(WDT(idx)), event_type,			       \
+				  requests, p_context);					       \
+	}										       \
+	), ())										       \
+	static int wdt_##idx##_init(const struct device *dev)				       \
+	{										       \
+		const struct wdt_nrfx_config *config = dev->config;			       \
+		nrfx_err_t err_code;							       \
+		COND_CODE_0(IS_ENABLED(CONFIG_WDT_NRFX_NO_IRQ), (			       \
+		IRQ_CONNECT(DT_IRQN(WDT(idx)), DT_IRQ(WDT(idx),				       \
+		priority), nrfx_isr, nrfx_wdt_##idx##_irq_handler, 0);), ())		       \
```

Comment:
This is quite hard to read. How about defining this in a macro:
```
#define WDT_NRFX_WDT_IRQ(idx)						       \
	COND_CODE_1(CONFIG_WDT_NRFX_NO_IRQ,				       \
		(),							       \
		(IRQ_CONNECT(DT_IRQN(WDT(idx)), DT_IRQ(WDT(idx), priority),    \
			     nrfx_isr, nrfx_wdt_##idx##_irq_handler, 0)))
```
and calling such macro here instead?